### PR TITLE
SONAR-8840 Create WS api/project_tags/search to search for tags

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SetActionTest.java [hunk 1]
```diff
@@ -51,20 +56,24 @@
   @Rule
   public DbTester db = DbTester.create();
   private DbClient dbClient = db.getDbClient();
-  public WsActionTester ws = new WsActionTester(new SetAction(dbClient, new ComponentFinder(dbClient), userSession));
   private DbSession dbSession = db.getSession();
   private ComponentDto project;
 
+  private ProjectIndexer indexer = mock(ProjectIndexer.class);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SetActionTest.java]: Unit tests of WS should not rely on mocks for DB/Index

File: server/sonar-server/src/main/java/org/sonar/server/projecttag/ws/SearchAction.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.server.measure.index.ProjectMeasuresIndex;
+import org.sonarqube.ws.WsProjectTags;
+
+import static org.sonar.api.server.ws.WebService.Param.PAGE_SIZE;
+import static org.sonar.api.server.ws.WebService.Param.TEXT_QUERY;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+
+public class SearchAction implements ProjectTagsWsAction {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projecttag/ws/SearchAction.java]: This class is missing in ProjectTagsWsModule => The WS is not available.

File: server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java [hunk 3]
```diff
@@ -304,6 +312,34 @@ private static QueryBuilder toValueQuery(MetricCriterion criterion) {
     }
   }
 
+  public List<String> searchTags(@Nullable String textQuery, int pageSize) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java]: Why only pageSize ? Either you should have page and pageSize, either you should return everything

File: server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SearchActionTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  @Test
+  public void json_example() throws IOException {
+    when(index.searchTags(anyString(), anyInt())).thenReturn(newArrayList("official", "offshore", "playoff"));
+
+    String result = ws.newRequest().execute().getInput();
+
+    assertJson(ws.getDef().responseExampleAsString()).withStrictArrayOrder().isSimilarTo(result);
+  }
+
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SearchActionTest.java]: Unit tests missing on some common use cases : search with text query, with page size, etc.

File: server/sonar-server/src/main/java/org/sonar/server/projecttag/ws/SearchAction.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projecttag/ws/SearchAction.java]: 😭

File: server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java]: it's the same for rules and issues tags. You cannot have pagination for an ES facet. That's what we're using here to return tags. Still, we want to be able to change the number of tags returned.

File: server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SearchActionTest.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SearchActionTest.java]: WDYM? Isn't it tested in ProjectMeasuresIndex? I don't want to test the same thing twice

File: server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SetActionTest.java (hunk 1 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/projecttag/ws/SetActionTest.java]: ok, main use case tested end to end

**Comment by teryk** [comment]
tickets for WS api/issues/tags and api/rules/tags have been created

File: server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java [hunk 5]
```diff
@@ -304,6 +312,34 @@ private static QueryBuilder toValueQuery(MetricCriterion criterion) {
...
+    checkArgument(pageSize <= 100, "Page size must be lower than or equals to " + 100);
+    if (pageSize == 0) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java]: Other WS on rules and issues tags do not have the same behavior => we should fix them (create JIRA tickets) or change this one

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java]: the ticket have already been created

**Pull request merged**